File: src/main/java/main/utils/database/mongodb/DocumentBuilder.java
package main.utils.database.mongodb;

import lombok.SneakyThrows;
import main.utils.component.InvalidBuilderExceptionKt;
import main.utils.json.GenericJSONField;
import org.bson.Document;
import org.json.JSONObject;

public class DocumentBuilder {
    private final JSONObject obj;

    private DocumentBuilder() {
        obj = new JSONObject();
    }

    public static DocumentBuilder create() {
        return new DocumentBuilder();
    }

    public <T> DocumentBuilder addField(String key, T value) {
        obj.put(key, value);
        return this;
    }

    public <T> DocumentBuilder addField(GenericJSONField key, T value) {
        obj.put(key.toString(), value);
        return this;
    }

    @SneakyThrows
    public Document build() {
        if (obj.isEmpty())
            throw new InvalidBuilderException("You cannot create a document with no fields!");

        return Document.parse(obj.toString());
    }
}

File: src/main/kotlin/main/utils/database/mongodb/DocumentBuilder.kt
package main.utils.database.mongodb

import main.utils.component.InvalidBuilderException
import main.utils.json.GenericJSONField
import org.bson.Document
import org.json.JSONObject

class DocumentBuilder private constructor() {
    private var obj: JSONObject = JSONObject()

    companion object {
        fun create(): DocumentBuilder = DocumentBuilder()
    }

    fun <T> addField(key: String, value: T): DocumentBuilder {
        obj.put(key, value)
        return this
    }

    fun <T> addField(key: GenericJSONField, value: T): DocumentBuilder {
        obj.put(key.toString(), value)
        return this
    }

    fun setObj(obj: JSONObject): DocumentBuilder {
        this.obj = obj
        return this
    }

    fun build(): Document = when {
        obj.isEmpty ->
            throw InvalidBuilderException("You cannot create a document with no fields!")
        else -> Document.parse(obj.toString())
    }

}